File: apps/desktop/src/native/api/project.ts
import type {
  ArrangementMutationResult,
  AssetId,
  HistoryState,
  ProjectActivationResult,
  ProjectState,
  ProjectExport,
} from '@/model/domain';
import { defaultProjectState, defaultSession } from '../browser-defaults';
import { invokeHostOrFallback, invokeHost } from '../invoke';

export async function undoSession(): Promise<ArrangementMutationResult> {
  const session = defaultSession();
  return invokeHostOrFallback<ArrangementMutationResult>(
    'undo_session',
    {},
    {
      canonical: {
        session,
        sequence: 0,
        history: { canUndo: false, canRedo: false },
      },
      projection: { state: 'notRequired' },
    },
  );
}

export async function redoSession(): Promise<ArrangementMutationResult> {
  const session = defaultSession();
  return invokeHostOrFallback<ArrangementMutationResult>(
    'redo_session',
    {},
    {
      canonical: {
        session,
        sequence: 0,
        history: { canUndo: false, canRedo: false },
      },
      projection: { state: 'notRequired' },
    },
  );
}

export async function getHistoryState(): Promise<HistoryState> {
  return invokeHostOrFallback<HistoryState>(
    'get_history_state',
    {},
    {
      canUndo: false,
      canRedo: false,
    },
  );
}

export async function restoreRecoveryGeneration(
  fileName: string,
): Promise<ArrangementMutationResult | null> {
  return invokeHostOrFallback<ArrangementMutationResult | null>(
    'restore_recovery_generation',
    { fileName },
    null,
  );
}

export async function exportProject(path: string): Promise<ProjectExport | null> {
  return invokeHostOrFallback<ProjectExport | null>('export_project', { path }, null);
}

export async function listProjects(): Promise<ProjectState> {
  return invokeHostOrFallback<ProjectState>('list_projects', {}, defaultProjectState());
}

export async function createProject(name?: string): Promise<ProjectActivationResult> {
  return invokeHostOrFallback<ProjectActivationResult>(
    'create_project',
    { name: name ?? null },
    defaultProjectActivationResult(),
  );
}

export async function openProject(projectId: string): Promise<ProjectActivationResult> {
  return invokeHostOrFallback<ProjectActivationResult>(
    'open_project',
    { projectId },
    defaultProjectActivationResult(),
  );
}

export async function renameProject(name: string): Promise<ProjectState> {
  return invokeHostOrFallback<ProjectState>('rename_project', { name }, defaultProjectState());
}

export async function importProject(path: string): Promise<ProjectActivationResult | null> {
  return invokeHostOrFallback<ProjectActivationResult | null>('import_project', { path }, null);
}

function defaultProjectActivationResult(): ProjectActivationResult {
  return {
    projectState: defaultProjectState(),
    canonical: {
      session: defaultSession(),
      sequence: 0,
      history: { canUndo: false, canRedo: false },
    },
    recovery: { recoveredFromGeneration: false, recoveryCandidates: [] },
  };
}

export async function importMidiFile(path: string, name?: string): Promise<AssetId | null> {
  return invokeHostOrFallback<AssetId | null>(
    'import_midi_file',
    { path, name: name ?? null },
    null,
  );
}

export async function importMidiBytes(name: string, bytes: number[]): Promise<AssetId | null> {
  return invokeHostOrFallback<AssetId | null>('import_midi_bytes', { name, bytes }, null);
}

export async function updateSessionSettings(patch: {
  projectName?: string | null;
  loopEnabled?: boolean;
  countInBeats?: number;
  metronomeEnabled?: boolean;
  note?: string;
}): Promise<ArrangementMutationResult> {
  const result = await invokeHost<ArrangementMutationResult>('update_session_settings', { patch });
  return result;
}

File: apps/desktop/src-tauri/src/session/commands/project.rs
use super::*;

use crate::model::ProjectState;

#[tauri::command]
pub async fn undo_session(app: AppHandle) -> Result<ArrangementMutationResult, String> {
    dispatch(app, "undo", json!({})).await
}

#[tauri::command]
pub async fn redo_session(app: AppHandle) -> Result<ArrangementMutationResult, String> {
    dispatch(app, "redo", json!({})).await
}

#[tauri::command]
pub async fn get_history_state(app: AppHandle) -> Result<HistoryState, String> {
    dispatch(app, "history.get", json!({})).await
}

#[tauri::command]
pub async fn restore_recovery_generation(
    file_name: String,
    app: AppHandle,
) -> Result<ArrangementMutationResult, String> {
    dispatch(
        app,
        "project.restore-generation",
        json!({ "fileName": file_name }),
    )
    .await
}

#[tauri::command]
pub async fn import_project(path: String, app: AppHandle) -> Result<ProjectState, String> {
    dispatch(app, "project.import", json!({ "path": path })).await
}

#[tauri::command]
pub async fn list_projects(app: AppHandle) -> Result<ProjectState, String> {
    dispatch(app, "project.list", json!({})).await
}

#[tauri::command]
pub async fn create_project(name: Option<String>, app: AppHandle) -> Result<ProjectState, String> {
    dispatch(app, "project.create", json!({ "name": name })).await
}

#[tauri::command]
pub async fn open_project(project_id: String, app: AppHandle) -> Result<ProjectState, String> {
    dispatch(app, "project.open", json!({ "projectId": project_id })).await
}

#[tauri::command]
pub async fn rename_project(name: String, app: AppHandle) -> Result<ProjectState, String> {
    dispatch(app, "project.rename", json!({ "name": name })).await
}

#[tauri::command]
pub async fn update_session_settings(
    patch: SessionSettingsPatch,
    app: AppHandle,
) -> Result<ArrangementMutationResult, String> {
    dispatch(app, "session.settings.update", patch).await
}

#[tauri::command]
pub async fn set_master_gain_db(gain_db: f64, app: AppHandle) -> Result<SessionAudioPair, String> {
    dispatch(app, "audio.master-gain.set", json!({ "gainDb": gain_db })).await
}

#[tauri::command]
pub async fn relink_missing_dependency(
    asset_id: String,
    new_path: String,
    app: AppHandle,
) -> Result<ArrangementMutationResult, String> {
    let asset_id = AssetId::from_normalized(asset_id)
        .map_err(|error| format!("Asset id is invalid: {error}"))?;
    dispatch(
        app,
        "missing.relink",
        json!({ "assetId": asset_id, "newPath": new_path }),
    )
    .await
}

#[tauri::command]
pub async fn disable_missing_plugin(
    device_id: String,
    app: AppHandle,
) -> Result<ArrangementMutationResult, String> {
    dispatch(
        app,
        "missing.disable-plugin",
        json!({ "deviceId": device_id }),
    )
    .await
}

#[tauri::command]
pub async fn replace_missing_track_plugin(
    device_id: String,
    new_path: String,
    app: AppHandle,
) -> Result<ArrangementMutationResult, String> {
    dispatch(
        app,
        "missing.replace-plugin",
        json!({ "deviceId": device_id, "newPath": new_path }),
    )
    .await
}

#[tauri::command]
pub async fn get_missing_dependencies(app: AppHandle) -> Result<Vec<MissingDependency>, String> {
    dispatch(app, "missing.list", json!({})).await
}
